### PR TITLE
add doc tracker retrieval action + refactor doc tracker actions setup

### PR DESCRIPTION
**Pull request by fontanierh**
- adds a new doc tracker retrieval action to the registry (unused as of now)
- refactor the setup to use actions in the doc tracker code: fully typed setup with response payload validation

### CODE REVIEW

**Review by spolu — DISMISSED** [review]
Feels "a bit complicated" but LGTM

**Comment by fontanierh** [comment]
> Feels "a bit complicated" but LGTM

@spolu 
Curious what specifically feels complicated. I think if we were to use `callAction` everywhere, then we could simply define a response value schema for each action and you automatically have typed responses instead of having to deal with `unknown`/`any`

**Review by lasryaric — DISMISSED** [review]
Same comment as @spolu . Feels a bit complicated, but looks like it's the price to pay to have typed parsing.

Thanks for exploring this path.

File: front/post_upsert_hooks/hooks/document_tracker/actions/index.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+interface CallActionParams<R extends ActionSuccessResponseBase> {
+  workspaceId: string;
+  input: unknown;
+  action: Action;
+  config: DustAppConfigType;
+  responseChecker: (response: unknown) => response is R;
+}
+
+async function callAction<R extends ActionSuccessResponseBase>({
```
Comment (spolu) [inline front/post_upsert_hooks/hooks/document_tracker/actions/index.ts:80]: Shouldn't we use a `Result` here?

What is `R["results"][0][0]["value"]` ?

**Comment by spolu** [comment]
Maybe it's because of the Legacy non Legacy stuff? It's generally hard to follow what's really going on :)

File: front/post_upsert_hooks/hooks/document_tracker/actions/index.ts (hunk 1 above)
Comment (lasryaric) [reply front/post_upsert_hooks/hooks/document_tracker/actions/index.ts:80]: +1

Comment (fontanierh) [reply front/post_upsert_hooks/hooks/document_tracker/actions/index.ts:80]: > What is R["results"][0][0]["value"]

It's the "value" part of the Dust API response

Comment (spolu) [reply front/post_upsert_hooks/hooks/document_tracker/actions/index.ts:80]: Yes yes but it's super hard to parse as a reader of this code without a ton of context :)

Comment (fontanierh) [reply front/post_upsert_hooks/hooks/document_tracker/actions/index.ts:80]: got it will try to make it clearer

File: front/post_upsert_hooks/hooks/document_tracker/actions/types.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  typeof DocTrackerLegacyActionSuccessResponseSchema
+>;
+export function isDocTrackerLegacyActionSuccessResponse(
+  response: unknown
+): response is DocTrackerLegacyActionSuccessResponse {
+  return isRight(DocTrackerLegacyActionSuccessResponseSchema.decode(response));
+}
+
+// Specific to the doc tracker retrieval action
+export const DocTrackerRetrievalActionValueSchema = t.type({
```
Comment (spolu) [inline front/post_upsert_hooks/hooks/document_tracker/actions/types.ts:86]: Should be removed (since it's in the other file?)

Comment (fontanierh) [reply front/post_upsert_hooks/hooks/document_tracker/actions/types.ts:86]: Forgot to delete thanks

File: front/post_upsert_hooks/hooks/document_tracker/actions/lib.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    }),
+  ]);
+  type responseType = t.TypeOf<typeof responseSchema>;
+  const responseChecker = (response: unknown): response is responseType =>
+    isRight(responseSchema.decode(response));
+
+  if (responseChecker(response.value)) {
+    // the response is a valid success response for the action
+    // return the "value" field of the first result
+    return response.value.results[0][0].value;
```
Comment (spolu) [inline front/post_upsert_hooks/hooks/document_tracker/actions/lib.ts:71]: Relies a lot on `results` also the fact that we have only one `input`... Maybe worth adding a comment that explains that for `callAction` ?

**Review by spolu — COMMENTED** [review]
Much clearer. Probably worth a small comment for `callAction` as it makes a few assumptions (one if enforced by the signature of the function (1 input)). The other assumes that we use only the last block of the app.

All that is fine but would be nice to explicit these assumptions clearly :+1:

File: front/lib/actions/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    // return the "value" field of the first result
+    return new Ok(response.value.results[0][0].value);
+  }
+
+  if (isActionResponseBase(response.value)) {
+    // the response is of the right shape, but it's not a success response
+    return new Err({
+      type: "action_failed",
+      message: `Doc Tracker action failed response: ${JSON.stringify(
+        response.value.status
```
Comment (spolu) [inline front/lib/actions/helpers.ts:92]: This is likely way tooo big ? (can be Mb)

File: front/lib/actions/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      )}`,
+    });
+  }
+
+  // the response is not of a known shape, so we can't assume anything about it
+  return new Err({
+    type: "unexpected_action_response",
+    message: `Unexpected Doc Tracker action response: ${JSON.stringify(
+      response.value
```
[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/actions/helpers.ts:101]: For this one I can't expect anything about `value`. It could be anything, so maybe we just don't want to log it ?

File: front/lib/actions/helpers.ts (hunk 4 above)
Comment (fontanierh) [reply front/lib/actions/helpers.ts:92]: Makes sense, for this one maybe I should dumb the keys + statuses ?

File: front/post_upsert_hooks/hooks/document_tracker/actions/lib.ts (hunk 3 above)
Comment (fontanierh) [reply front/post_upsert_hooks/hooks/document_tracker/actions/lib.ts]: > Relies a lot on results

What do you mean exactly by this ? `results` is always the output of the last block ?

Comment (fontanierh) [reply front/post_upsert_hooks/hooks/document_tracker/actions/lib.ts]: > Maybe worth adding a comment that explains that for callAction ?

agreed, adding

File: front/lib/actions/helpers.ts (hunk 4 above)
Comment (spolu) [reply front/lib/actions/helpers.ts:92]: Yes anything what does not have the traces or results :)

File: front/lib/actions/helpers.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (spolu) [reply front/lib/actions/helpers.ts:101]: I think that's fine not logging it and revisiting if needed :+1:

File: front/lib/actions/helpers.ts (hunk 4 above)
Comment (fontanierh) [reply front/lib/actions/helpers.ts:92]: How can this be megabytes ? It's not the trace, just the status for each block no ? 
Screenshot 2023-07-17 at 15 25 01

should be this shape if I'm not mistaken ?

Comment (spolu) [reply front/lib/actions/helpers.ts:92]: Ah sorry I missed that it was response.value.status thought it was response.value. reponse.value.status is fine :+1:

**Review by spolu — APPROVED** [review]
